Update overview documentation (#63)
MicrosoftDocs/microsoft-authentication-library-for-python: msal-python-conceptual/index.md

---
title: Overview of the Microsoft Authentication Library (MSAL) for Python
description: Get started with the Microsoft Authentication Library for Python to sign in users or apps with Microsoft identities."
author: SHERMANOUKO
manager: CelesteDG

ms.service: msal
ms.subservice: msal-python
ms.topic: conceptual
ms.date: 02/29/2024
ms.author: shermanouko
ms.reviewer: dmwendia, rayluo
---

# Microsoft Authentication Library (MSAL) for Python

> The Microsoft Authentication Library (MSAL) for Python library enables you to sign in users or apps with Microsoft identities. These identities include work and school accounts, personal microsoft accounts, social accounts, customer accounts and so on. Using MSAL Python, you can acquire tokens from Microsoft Entra to call protected web APIs such as [Microsoft Graph](https://graph.microsoft.io/), other Microsoft APIs, or your own APIs. MSAL Python supports a variety of application types, including public client applications (desktop and mobile) and confidential client applications (web apps, web APIs, and daemon applications).

Reviewer: ```suggestion
The Microsoft Authentication Library (MSAL) for Python library enables you to sign in users or apps with Microsoft identities. These identities include work and school accounts, personal Microsoft accounts, social accounts, customer accounts, and so on. Using MSAL Python, you can acquire tokens from Microsoft Entra to call protected web APIs such as [Microsoft Graph](https://graph.microsoft.io/), other Microsoft APIs, or your own APIs. MSAL Python supports various application types, including public client applications (desktop and mobile) and confidential client applications (web apps, web APIs, and daemon applications).
```

## Identity concepts

MSAL Python is part of the [Microsoft identity platform](/entra/identity-platform/v2-overview) ecosystem. It's important to farmiliarize yourself with the following concepts to effectively use MSAL Python to protect your applications and APIs:

- [Identity and access management](/entra/fundamentals/identity-fundamental-concepts)
- [Authentication and authorization](/entra/identity-platform/authentication-vs-authorization)
- [OAuth 2.0 and OpenID Connect (OIDC) in the Microsoft identity platform](/entra/identity-platform/v2-protocols)
- [Confidential and public client accounts in the Microsoft identity platform](/entra/identity-platform/msal-client-applications)
- [Security tokens](/entra/identity-platform/security-tokens)

## Usage scenarios

To use MSAL Python, register an application with the Microsoft identity platform. You'll need an Azure account with an active subscription. [Create a free account](https://signup.azure.com/) if you don't have one. You can register your app in a [customer tenant](/entra/external-id/customers/quickstart-tenant-setup) or [workforce tenant](/entra/identity-platform/scenario-web-app-sign-user-app-registration?tabs=python).

> MSAL Python can be used by applications to acquire tokens to access protected APIs. Tokens can be acquired by different application types. These app types include desktop applications, web applications, web APIs, and applications running on devices that don't have a browser (such as IoT devices). Different app types follow different auth flows.

Reviewer: ```suggestion
Applications can use MSAL Python to acquire tokens for accessing protected APIs. Different app types acquire tokens using different auth flows. The supported app types include desktop applications, web applications, web APIs, and applications running on devices that don't have a browser (such as IoT devices). 
```

Reviewer: Not sure if this bit sounds any better, but the idea is to eliminate the passive voice in two consecutive sentences.

1. Try and obtain the tokens from the cache first. The API model in MSAL provides you explicit control on how to utilize the token cache. While the caching part is technically optional, we highly recommend you to use it in your application. Using the cache you can ensure that you're not making any extra API calls and handle the token refresh automatically.

    ```python
    # initialize result variable to hole the token response
    result = None 
    
    # We now check the cache to see
    # whether we already have some accounts that the end user already used to sign in before.
    accounts = app.get_accounts()
    if accounts:
        # If so, you could then somehow display these accounts and let end user choose
        print("Pick the account you want to use to proceed:")
        for a in accounts:
            print(a["username"])
        # Assuming the end user chose this one
        chosen = accounts[0]
        # Now let's try to find a token in cache for this account
        result = app.acquire_token_silent(["User.Read"], account=chosen)
    ```

> 1. If there is no suitable token in the cache or you've chosen to skip the previous step, send a request to Microsoft Entra ID to get a token. There are different methods based on your client type and scenario, but for the purposes of the example we're showing how to use [`acquire_token_interactive`](xref:msal.application.PublicClientApplication.acquire_token_interactive) which will prompt the user to provide their credentials.

Reviewer: ```suggestion
1. If there's no suitable token in the cache or you chose to skip the previous step, send a request to Microsoft Entra ID to get a token. There are different methods based on your client type and scenario, but for the purposes of the example we're showing how to use [`acquire_token_interactive`](xref:msal.application.PublicClientApplication.acquire_token_interactive), which prompts the user to provide their credentials.
```

       print(result.get("correlation_id"))  # You may need this when reporting a bug
   ```

1. Save the code into a Python file locally, such as `msaltest.py`.
1. Run the code by executing `python .\msalpytest.py`. The following visual shows the sign-in experience for this example.

    ![Example of an app prompting the user to sign in with their account](./media/basic-pca-app-prompt.gif)

1. Once the authentication is completed and you closed the browser, you should be able to see the access token printed in the terminal.

## Samples

> TRhere are several samples you can use to get started with MSAL Python.

Reviewer: ```suggestion
There are several samples you can use to get started with MSAL Python.
```

Reviewer: minor typo here

> - Samples from the [library repository](https://github.com/AzureAD/microsoft-authentication-library-for-python/blob/1.22.0/sample/interactive_sample.py).

Reviewer: ```suggestion
- Samples from the [library repository](https://github.com/AzureAD/microsoft-authentication-library-for-python/blob/1.22.0/sample). These samples demonstrate the different configurations and auth flows that you implement using MSAL Python.
```

Reviewer: Let's point to the samples directory here, rather than the first code sample.
